provisioning/service/samples/GroupCertificateVerificationSample/VerificationCertificateGenerator.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Security.Cryptography.X509Certificates;
+using Org.BouncyCastle.Asn1;
+using Org.BouncyCastle.Asn1.X509;
+using Org.BouncyCastle.Crypto;
+using Org.BouncyCastle.Crypto.Generators;
+using Org.BouncyCastle.Crypto.Operators;
+using Org.BouncyCastle.Crypto.Parameters;
+using Org.BouncyCastle.Crypto.Prng;
+using Org.BouncyCastle.Math;
+using Org.BouncyCastle.Pkcs;
+using Org.BouncyCastle.Security;
+using Org.BouncyCastle.X509;
+
+namespace Microsoft.Azure.Devices.Provisioning.Samples
+{
+    public class VerificationCertificateGenerator
+    {
+        public static X509Certificate2 GenerateSignedCertificate(X509Certificate2 signerCertificate, string commonName)
+        {
+            string subjectName = string.Concat("CN=", commonName);
+            var name = new X509Name(subjectName);
+            X509Name signerName = DotNetUtilities.FromX509Certificate(signerCertificate).SubjectDN;
+
+            Pkcs12Store store = CreateTemporaryStore(signerCertificate);
+            ECPrivateKeyParameters privateKey = GetPrivateKey(store);
+            X509Certificate2 certificate = GenerateAndSignCertificate(name, signerName, privateKey);
+            return certificate;
+        }
+
+        private static Pkcs12Store CreateTemporaryStore(X509Certificate2 signerCertificate)
+        {
+            Pkcs12Store store = new Pkcs12StoreBuilder().Build();
+
+            var random = new SecureRandom();
+            string sessionPassword = Convert.ToBase64String(SecureRandom.GetNextBytes(random, 32));
+
+            using (var stream = new MemoryStream(signerCertificate.Export(X509ContentType.Pfx, sessionPassword)))
+            {
+                store.Load(stream, sessionPassword.ToCharArray());
+            }
+
+            return store;
+        }
+
+        private static ECPrivateKeyParameters GetPrivateKey(Pkcs12Store store)
+        {
+            ECPrivateKeyParameters privateParameter = null;
+
+            foreach (string alias in store.Aliases)
+            {
+                var key = store.GetKey(alias);
+
+                if (key != null && key.Key.IsPrivate)
+                {
+                    privateParameter = key.Key as ECPrivateKeyParameters;
+                }
+            }
+
+            return privateParameter;
+        }
+
+        private static X509Certificate2 GenerateAndSignCertificate(
```

Comment:
A new requirement I know but we should actually break this into generating a CSR and consuming it. This is how it really should happen and makes this extensible to using an actual CSR service.

Reply:
While that may be the way a real CA or Intermediate CA would do it, this is just providing a way for the developer to bootstrap their development process. 

On Windows, the methods I'm aware of are via a [wizard](https://technet.microsoft.com/en-us/library/cc754490(v=ws.11).aspx) or via [certreq](https://technet.microsoft.com/en-us/library/dn296456(v=ws.11).aspx).

Reply:
This serves the dev bootstrap scenario well, we can produce CSR based samples as a more formal way of doing this.

Reply:
I'll start an internal thread to have this added to the documentation on OSs that we support.

Reply:
@yzhong94 PTAL